File: C++/1330-Reverse Subarray To Maximize Array Value.cpp
class Solution {
public:
    int maxValueAfterReverse(vector<int>& nums) {
        int mx = INT_MIN, mn = INT_MAX;
        int change = 0;
        int N = nums.size();
        int tot = 0;
        for (int i=0; i<N-1; i++){
            int a = nums[i], b = nums[i+1];
            tot += abs(a-b);
            mx = max(min(a, b), mx);
            mn = min(max(a, b), mn);
        }
        change = max(change, 2*(mx-mn));
        
        for (int i=0; i<N-1; i++){
            int a = nums[i], b = nums[i+1];
            int tmp1 = -abs(a-b) + abs(nums[0]-b);
            int tmp2 = -abs(a-b) + abs(a-nums[N-1]);
            change = max(change, max(tmp1, tmp2));
        }       

        return tot + change;

    }
};

File: 1330-Reverse Subarray To Maximize Array Value.py
class Solution:
    def maxValueAfterReverse(self, nums: List[int]) -> int:
# https://leetcode.com/problems/reverse-subarray-to-maximize-array-value/discuss/489882/O(n)-Solution-with-explanation 
        mx, mn = -float('inf'), float('inf')
        change = 0
        for a, b in zip(nums, nums[1:]):
            mx = max(min(a, b), mx)
            mn = min(max(a, b), mn)
            change = max(change, 2*(mx-mn))
        
        for a, b in zip(nums, nums[1:]):
            tmp1 = -abs(a-b) + abs(nums[0] - b)
            tmp2 = -abs(a-b) + abs(a - nums[-1])
            change = max(change, tmp1, tmp2)
        
        tot = sum(abs(a-b) for a, b in zip(nums, nums[1:]))
        
        return tot + change